src/System/Interactions.jl

```diff
@@ -312,73 +305,70 @@ function set_forces_aux!(B, dipoles::Array{Vec3, 4}, ints::Interactions, sys::Sy
     if N == 0
         for cell in cells
             s = dipoles[cell, i]
-            B[cell, i] -= energy_and_gradient_for_classical_anisotropy(s, ints.aniso.stvexp)[2]
+            B[cell, i] -= energy_and_gradient_for_classical_anisotropy(s, ints.onsite.stvexp)[2]
         end
     end
 
-    # Heisenberg exchange
-    foreachbond(ints.heisen) do J, site1, site2
-        sᵢ = dipoles[site1]
-        sⱼ = dipoles[site2]
-        B[site1] -= J  * sⱼ
-        B[site2] -= J' * sᵢ
-    end
-
-    # Quadratic exchange matrix
-    foreachbond(ints.exchange) do J, site1, site2
+    foreachbond(ints.pair) do coupling, site1, site2
         sᵢ = dipoles[site1]
         sⱼ = dipoles[site2]
-        B[site1] -= J  * sⱼ
-        B[site2] -= J' * sᵢ
-    end
 
-    # Scalar biquadratic exchange
-    foreachbond(ints.biquad) do J, site1, site2
-        sᵢ = dipoles[site1]
-        sⱼ = dipoles[site2]
+        if coupling.bilin isa Float64
+            J = coupling.bilin::Float64
+            B[site1] -= J  * sⱼ
+            B[site2] -= J' * sᵢ
+        else
+            J = coupling.bilin::Mat3
+            B[site1] -= J  * sⱼ
+            B[site2] -= J' * sᵢ
+        end
```

Comment:
Ditto, these code are the same thing and is type stable, don't branch explicitly

Reply:
OH, I see you're restricting the possible options it has to deal with. I think it's probably still not needed...

Reply:
Another possibility: This is what `UniformScaling` was designed for, right? there's also `dot(v,::UniformScaling,v)`

Reply:
This code was the result of some experiments and reading the docs. Maybe easiest to discuss on slack.

Reply:
I have updated the PR to use multiplication of `Union{Float64,Mat3}`, which Julia seems to optimize well. Also, if there were type instability, the JET.jl unit test seem to be good about finding it.